test/self-destructor.ts

```diff
@@ -0,0 +1,123 @@
+   */
+  selfDestructAfterSeconds?: number;
+}
+
+export class SelfDestructor extends Construct {
+  constructor(scope: Construct, id: string, props?: SelfDestructorProps) {
+    super(scope, id);
+
+    const stack = Stack.of(this);
+    // arn:aws:cloudformation:us-east-1:123456789012:stack/MyProductionStack/*
+    const stackArn = stack.formatArn({
+      resource: "stack",
+      service: "cloudformation",
+      resourceName: stack.stackName,
+    });
+
+    /**
+     * It is not possible to do timestamp math in StepFunctions unfortunately.
+     *
+     * Returns `true` if `timestamp + offset` is before the current date (`Date.now()`).
+     */
+    const dateAddSeconds = new Function(
+      this,
+      "timestampOffsetBeforeCurrent",
+      async (input: {
+        timestamp: Date | string;
+        offsetSeconds: number;
+      }): Promise<string> => {
+        const date = new Date(input.timestamp);
+        date.setSeconds(date.getSeconds() + input.offsetSeconds);
+        return date.toISOString();
+      }
+    );
+
+    const selfDestructMachine = new StepFunction(
+      this,
+      "selfDestruct",
+      async (input: SelfDestructorMachineProps) => {
+        let destructTime = "";
+        do {
+          const stackResult = await $AWS.SDK.CloudFormation.describeStacks(
+            {
+              StackName: stack.stackName,
+            },
+            {
+              iam: {
+                resources: [`${stackArn}/*`],
+              },
+            }
+          );
+
+          const currentStack = stackResult.Stacks?.[0] ?? null;
+          if (!currentStack) {
+            throw new StepFunctionError("StackNotFound", "Stack was not found");
+          }
+
+          const stackChangeTime = (currentStack.LastUpdatedTime ??
+            currentStack.CreationTime) as unknown as string;
+
+          // if we made it through a wait period and stackChangeTime is still the same the last iteration, destroy the stack
+          if (destructTime === stackChangeTime) {
+            /**
+             * Stack deletion happens asynchronously after this call succeeds.
+             */
+            await $AWS.SDK.CloudFormation.deleteStack(
+              {
+                StackName: stack.stackName,
+              },
+              {
+                iam: { resources: [`${stackArn}/*`] },
+              }
+            );
+            return null;
+          }
+
+          // add the destruct delay to the last update or creation time.
+          const offsetTime = await dateAddSeconds({
+            timestamp: stackChangeTime,
+            offsetSeconds: input.selfDestructAfterSeconds ?? 43200,
+          });
+
+          // set the base time to the destruct time, if these are the same after selfDestructAfterSeconds, destroy the stack
+          destructTime = stackChangeTime as unknown as string;
+
+          // wait until selfDestructAfterSeconds + stackChangeTime
+          $SFN.waitUntil(offsetTime);
```

Comment:
I still wish this was an await. It's always jarring

Reply:
though could be convinced otherwise...

Durable functions uses await on timers.

https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-timers?tabs=csharp#usage-for-delay

So does temporal

https://docs.temporal.io/workflows/#intrinsic-non-deterministic-logic